src/renderer/store/selection/reducer.ts

```diff
@@ -0,0 +1,71 @@
+import { Reducer } from 'redux';
+import { IndexRange } from '../../../sharedTypes';
+import {
+  SELECTION_RANGE_ADDED,
+  SELECTION_RANGE_REMOVED,
+  SELECTION_CLEARED,
+  SELECTION_RANGE_TOGGLED,
+} from './actions';
+import { ApplicationStore, initialStore } from '../sharedHelpers';
+import { Action } from '../action';
+
+const selectionReducer: Reducer<ApplicationStore['selection'], Action<any>> = (
+  selection = initialStore.selection,
+  action
+) => {
+  if (action.type === SELECTION_RANGE_ADDED) {
+    const { startIndex, endIndex } = action.payload as IndexRange;
+
+    /**
+     * Selection is stored in redux as an array because redux
+     * is only supposed to store primitives. So cast it to a set to manipulate then
+     * convert it back. O(n) with respect to length of transcription which is the same as
+     * the degenerate "select everything" case anyway
+     */
+    const selectionSet = new Set(selection);
```

Comment:
Love your work.

Is it possible to put this functionality in a function? That function could take a modification function as a parameter (eg: `add()` or `delete()`) or maybe an enumerator value (eg: `'ADD'` for `add()`). Could be put in a helpers file or something.

Just want to make it a bit easier to read by making it more compact. :)

Reply:
Put what in a function specifically? The entire reducer for SELECTION_RANGE_ADDED? It's pretty short, I'm not sure what the benefit would be vs keeping the selection manipulation reducers in one place

Reply:
Maybe - what about if it was

```
enum SelectionModifier {
  ADD,
  REMOVE,
  TOGGLE
}

const updateSelection: (selection: number[], modifier: SelectionModifier) => number[];
```

And then just internally

```
const mapper: Record<SelectionModifier, (wasSelected: boolean) => boolean> = {
  SelectionModifier.ADD: () => true,
  SelectionModifier.REMOVE: () => false,
  SelectionModifier.TOGGLE: (wasSelected) => !wasSelected,
}
```

Oops I just 3/4 implemented it in this comment 🙈

Reply:
To clarify, I meant the logic from put selection in set, for loop and turn back to array in a separate function (with the for loop modifier as a parameter). Just an idea.

All good if you don't think its worth it.

Reply:
Sorry, confused by comments. If it reduces the re-used lines then yes please! :)

Reply:
Yep that’s what the proposed impl does :)